feat(core): move ANIMATION_MODULE_TYPE injection token into core

crisbeto opened the pull request:
Moves the `ANIMATION_MODULE_TYPE` DI token into core so that libraries like Material don't have to introduce a dependency to animations only to figure out whether animations are disabled.

AndrewKushnir reviewed (commented):
@crisbeto the change looks good, but I'd like to understand the motivation for this change a bit better. Are we trying to avoid relying on the `@angular/animations` in some components to reduce the build time (so there is no need to build that package) or there are other reasons?

Just trying to understand the use-case a bit more before we introduce an animations-specific token in the core package as a public API.

crisbeto commented:
In Angular Material some components are animated through Angular animations while others are animated directly with CSS. There were reports that it was inconsistent when including the `NoopAnimationsModule` disables the Angular-based animations, but not the CSS-based ones. It also caused flakes in screenshot tests, because users didn't have a single point at which to disable animations. This DI token was introduced as a way to indicate that the user opted out of animations completely.

This PR moves the token into core as discussed, because while we can get away with using it on Material, because we already have a dependency on the animations package, we can't use it in the CDK without introducing another dependency.

AndrewKushnir commented on packages/core/src/application_tokens.ts:
```diff
@@ -72,3 +72,11 @@ export const APP_BOOTSTRAP_LISTENER =
+ * module has been loaded.
+ * @publicApi
+ */
+export const ANIMATION_MODULE_TYPE =
```
Could you please add a bit more context here on **why** we have this token as a part of the `@angular/core`? This comment could be outside of the public docs, just an internal comment in the code (so it's not visible on AIO).

AndrewKushnir replied:
Also it'd be great to include that into the commit message.

crisbeto replied:
I've added a comment in the code as well, but it seems a little redundant to repeat the same information in code, in the commit and the discussion of this PR.

AndrewKushnir replied:
I wanted to make sure this context is easily discoverable in the future, so extra redundancy is ok :)

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

crisbeto commented:
Capturing what we chatted about: technically there's nothing stopping us from importing the token from `@angular/platform-browser/animations` in the CDK, however we don't want to do that, because it would introduce an extra dependency. Currently the CDK doesn't reference `platform-browser` at all, but if we were to import the token from there, users would have to install `@angular/platform-browser` and potentially `@angular/animations` which `platform-browser` has a peer dependency on. Moving the token allows us to keep avoid requiring extra packages to be installed. Furthermore, there's already precedent for animation-specific code in `core` in the form of the `animations` field in the `@Component` metadata, as well as the animations-specific instructions in Ivy.

atscott commented:
Hey @crisbeto, isn't this a breaking change? I don't think we necessarily need a migration since it's easy enough to find/replace the import.

crisbeto commented:
It's not a breaking change, because the symbol is still re-exported through the old entry point.

atscott commented:
> It's not a breaking change, because the symbol is still re-exported through the old entry point.

Ah, I see that now. LGTM

atscott reviewed (approved):
reviewed-for: public-api

jessicajaniuk reviewed (approved):
LGTM :cookie:

Also curious about @AndrewKushnir's question.

reviewed-for: public-api, fw-animations, size-tracking

atscott reviewed (approved):
reviewed-for: size-tracking

crisbeto committed:
feat(core): move ANIMATION_MODULE_TYPE injection token into core

Moves the `ANIMATION_MODULE_TYPE` DI token into core so that libraries like Material don't have to introduce a dependency to animations only to figure out whether animations are disabled.

alxhub commented:
This PR was merged into the repository by commit e461f716d4b736829d53de26ba6dddb58f476746.
